### PR TITLE
solana-validator: 1.10 -> 1.16

### PR DESCRIPTION
**Pull request by polachok**
## Description of changes

Updated to latest stable version

## Things done

* fixed build on darwin inspired by blockchains/solana
* fixed build with in tree rust with a patch (basically same way as in master)

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
 I have only tested `solana-test-validator`, which seems to work fine
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/blockchains/solana-validator/default.nix
Comment (a-n-n-a-l-e-e) [inline pkgs/applications/blockchains/solana-validator/default.nix:63]: upstream patches using [fetchpatch](https://nixos.org/manual/nixpkgs/stable/#fetchpatch) are preferred, looks like this change is contained in the following commit, https://github.com/solana-labs/solana/commit/9e703f85de4184f577f22a1c72a0d33612f2feb1

```nix
(fetchpatch {
  url = "https://github.com/solana-labs/solana/commit/9e703f85de4184f577f22a1c72a0d33612f2feb1.patch";
  hash = "";
  includes = [ "path/to/account_info.rs" ];
})
```

File: pkgs/applications/blockchains/solana-validator/default.nix
Comment (a-n-n-a-l-e-e) [inline pkgs/applications/blockchains/solana-validator/default.nix:97]: env variables should be under the `env` attribute, eg:

```nix
  env = {
    VAR1 = "abc";
    VAR2 = "efg";
};
```

**Review by ghost — COMMENTED** [review]
looks pretty good, just a couple of comments.

File: pkgs/applications/blockchains/solana-validator/default.nix
Comment (polachok) [reply pkgs/applications/blockchains/solana-validator/default.nix:63]: what do I do if it doesnt apply cleanly?

Comment (a-n-n-a-l-e-e) [reply pkgs/applications/blockchains/solana-validator/default.nix:63]:
```nix
  patches = [
    (fetchpatch {
       url = "https://github.com/solana-labs/solana/commit/9e703f85de4184f577f22a1c72a0d33612f2feb1.patch";
       hash = "sha256-bAKTIQ6FhTk6bIddYULwLfdH5kzNPw1ltXJEfawtAXg=";
       includes = [ "sdk/program/src/account_info.rs" ];
     })
  ];
```
applies cleanly for me.

Comment (a-n-n-a-l-e-e) [reply pkgs/applications/blockchains/solana-validator/default.nix:63]: if fetchpatch seems to be doing the wrong thing, it might be because the hash is stale. if you modify includes, you also need to reset the hash otherwise you'll get stale data.

Comment (polachok) [reply pkgs/applications/blockchains/solana-validator/default.nix]: Yeah, sorry I missed the `includes` part

File: pkgs/applications/blockchains/solana-validator/default.nix
Comment (a-n-n-a-l-e-e) [inline pkgs/applications/blockchains/solana-validator/default.nix:65]: just need a comment for the patch.

**Comment by a-n-n-a-l-e-e** [comment]
change looks fine to me but just noticed that this is similar to https://github.com/NixOS/nixpkgs/blob/ce19e23bc46e86818764f44df41d1d973132bfc9/pkgs/applications/blockchains/solana/default.nix so cc solana maintainers just in case they have any comments.
cc solana maintainers @netfox @happysalada @aikooo7

**Comment by a-n-n-a-l-e-e** [comment]
Result of `nixpkgs-review pr 280118` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>solana-validator</li>
  </ul>
</details>

---

Result of `nixpkgs-review pr 280118` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>solana-validator</li>
  </ul>
</details>

File: pkgs/applications/blockchains/solana-validator/default.nix
Comment (aikooo7) [inline pkgs/applications/blockchains/solana-validator/default.nix:49]: Unused bind

File: pkgs/applications/blockchains/solana-validator/default.nix
Comment (aikooo7) [inline pkgs/applications/blockchains/solana-validator/default.nix:64]: Wouldn't be using ``name = "rustc-1.7x.patch";`` better?

**Review by aikooo7 — APPROVED** [review]
Regardless of the nit picks, everything looks fine for me

**Review by ghost — APPROVED** [review]

File: pkgs/applications/blockchains/solana-validator/default.nix
Comment (a-n-n-a-l-e-e) [reply pkgs/applications/blockchains/solana-validator/default.nix:64]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-updates says a comment should accompany a patch. I like having a self-documenting name but I've been dinged by reviewers for missing a comment so usually just add both to expedite reviews but will accept either in changes i review.

File: pkgs/applications/blockchains/solana-validator/default.nix
Comment (a-n-n-a-l-e-e) [reply pkgs/applications/blockchains/solana-validator/default.nix:49]: nice catch. this line should be removed then i will merge this unless there are any objections.

**Review by ghost — APPROVED** [review]

**Pull request merged**
